File: scripts/generate_index.js
const fs = require('fs');
const path = require('path');

(async function(){
  try {
    const postsDir = path.join(__dirname, '..', 'docs', 'posts');
    const outFile = path.join(__dirname, '..', 'docs', 'index.html');

    const files = await fs.promises.readdir(postsDir);
    const posts = [];

    for (const f of files) {
      if (!f.toLowerCase().endsWith('.html')) continue;
      const full = path.join(postsDir, f);
      const content = await fs.promises.readFile(full, 'utf8');

      // Extract title
      const titleMatch = content.match(/<h1[^>]*class=["']post-title["'][^>]*>([\s\S]*?)<\/h1>/i);
      const title = titleMatch ? titleMatch[1].trim() : f;

      // Extract datetime
      const timeMatch = content.match(/<time[^>]*class=["']post-date["'][^>]*datetime=["']([^"']+)["'][^>]*>([\s\S]*?)<\/time>/i);
      const datetime = timeMatch ? timeMatch[1].trim() : null;
      const displayDate = timeMatch ? timeMatch[2].trim() : (datetime || (await fs.promises.stat(full)).mtime.toISOString().slice(0,10));

      posts.push({ file: f, title, datetime: datetime || (await fs.promises.stat(full)).mtime.toISOString(), displayDate });
    }

    posts.sort((a,b) => (a.datetime < b.datetime ? 1 : -1));

    const listItems = posts.map(p => `  <li><a href="posts/${p.file}">${p.title}</a> <small>${p.displayDate}</small></li>`).join('\n');

    const html = `<!doctype html>
<html lang="ja">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <title>ブログ記事一覧</title>
</head>
<body>
  <h1>ブログ記事一覧</h1>
  <ul>
${listItems}
  </ul>
</body>
</html>`;

    await fs.promises.writeFile(outFile, html, 'utf8');
    console.log('Generated', outFile);
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
})();

File: scripts/generate_index.py
#!/usr/bin/env python3
import os
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
POSTS_DIR = ROOT / 'docs' / 'posts'
OUT_FILE = ROOT / 'docs' / 'index.html'

title_re = re.compile(r'<h1[^>]*class=["\']post-title["\'][^>]*>([\s\S]*?)</h1>', re.I)
time_re = re.compile(r'<time[^>]*class=["\']post-date["\'][^>]*datetime=["\']([^"\']+)["\'][^>]*>([\s\S]*?)</time>', re.I)

posts = []
for p in POSTS_DIR.iterdir():
    if p.suffix.lower() != '.html' or not p.is_file():
        continue
    text = p.read_text(encoding='utf-8')
    m = title_re.search(text)
    title = m.group(1).strip() if m else p.stem
    tm = time_re.search(text)
    if tm:
        datetime = tm.group(1).strip()
        display = tm.group(2).strip()
    else:
        stat = p.stat()
        datetime = stat.st_mtime
        # fallback ISO date
        from datetime import datetime as _dt
        display = _dt.fromtimestamp(stat.st_mtime).strftime('%Y-%m-%d')
        datetime = _dt.fromtimestamp(stat.st_mtime).isoformat()
    posts.append({'file': p.name, 'title': title, 'datetime': datetime, 'display': display})

posts.sort(key=lambda x: x['datetime'], reverse=True)

items = []
for it in posts:
    items.append(f'  <li><a href="posts/{it["file"]}">{it["title"]}</a> <small>{it["display"]}</small></li>')

html = f"""<!doctype html>
<html lang="ja">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <title>ブログ記事一覧</title>
</head>
<body>
  <h1>ブログ記事一覧</h1>
  <ul>
{os.linesep.join(items)}
  </ul>
</body>
</html>
"""

OUT_FILE.write_text(html, encoding='utf-8')
print('Generated', OUT_FILE)
